### PR TITLE
Categorical TOC model and src folder organization

### PR DESCRIPTION
**Pull request by SamStruthers**
Hi B and Juan!

I finished my first pass at a categorical TOC model, which more or less matches the same flow/formatting as the `03_ross_only_TOC` script. The primary differences are in how we assess performance (visually and numerically). There are a few new plotting functions to create confusion matrix plots and there is a bit of difference in how predictions are calculated based on using `multi:softmax` or `multi: softprob` in xgboost. While it was a bit more annoying to deal with, I ended up using `multi:softprob` (which outputs the class probabilities) so that we could do soft voting on ensemble predictions. I checked the differences between the voting methods and they are not that different but the ML textbook we have been reading seems to suggest that they will get better as we add more data/ensemble members. On our test set we get a model accuracy of 89% which is pretty good! Similar to linear regression, we don't do as well on the super high category. 

I also reorganized the src folder to close out #62 . I updated the files I commonly use @steeleb if you could update your files when you push your PR that would be great!

#Review/Thoughts:
I am wondering if we should add an additional TOC class splitting 4-8 into 4-6 and 6-8. Our accuracy will decrease but just wondering if it would be a more useful tool for our partners? @steeleb probably something we can ask on Tuesday

I split out these plotting functions from the regression plots but I am curious if we want to make a more general function that will decide if it is a categorical or regression model and make a plot accordingly?

@juandlt-csu please double check that I didn't miss anything in the `src/data_collation` scripts that will break GH Actions when I merge. 

@steeleb Any feedback on this workflow is always appreciated! Especially anything relating to performance metrics we want to be thinking about would be helpful! I added a few that seemed to come up again and again in my learning (accuracy, F1, kappa) but lmk if I am missing anything big!

### CODE REVIEW

File: modeling/toc/03_ross_only_categorical_TOC_model.Rmd [hunk 1]
```diff
@@ -0,0 +1,1281 @@
+---
+title: "Testing TOC preds"
+author: "Sam Struthers-ROSSyndicate"
+date: "2025-08-01"
+output: html_document
+---
+
```
Comment (steeleb) [inline modeling/toc/03_ross_only_categorical_TOC_model.Rmd]: suggestions: add a quick purpose statement here

File: modeling/toc/03_ross_only_categorical_TOC_model.Rmd [hunk 2]
```diff
@@ -0,0 +1,1281 @@
...
+  tail(1)
+
+# Read in all matches data
+all_matches <- read_csv(file = matchup_dataset, show_col_types = F) %>%
+  apply_sensor_transformations_toc(dt_col = "sensor_datetime")%>%
+  #remove NAs
+  filter(!is.na(TOC))%>%
+  mutate(clean_sensor_datetime = parse_date_time(sensor_datetime, orders = c("%Y-%m-%d %H:%M:%S", "%m/%d/%y %H:%M","%Y-%m-%d" )))%>%
+  mutate(sensor_datetime = force_tz(clean_sensor_datetime, tzone = 'MST'), 
+         month = lubridate::month(sensor_datetime),
```
Comment (steeleb) [inline modeling/toc/03_ross_only_categorical_TOC_model.Rmd]:
```suggestion
         month = month(sensor_datetime),
```

File: modeling/toc/03_ross_only_categorical_TOC_model.Rmd [hunk 3]
```diff
@@ -0,0 +1,1281 @@
...
+  )
+
+```
+
+## Define ID Columns, Target, and Features
+
+```{r}
+# ID columns to keep to use later on 
+id_cols <- c("Date", "id","site", "DT_hourly_round", "sensor_datetime","lab_datetime",  "collector", "data_avail")
+#What we want to predict
```
Comment (steeleb) [inline modeling/toc/03_ross_only_categorical_TOC_model.Rmd]:
```suggestion
# What we want to predict
```

File: modeling/toc/03_ross_only_categorical_TOC_model.Rmd [hunk 4]
```diff
@@ -0,0 +1,1281 @@
...
+  #These are not applicable sites for modeling Upper CLP streams and will just introduce error
+  filter( id %nin% c("springcreek", "boxcreek"))
+# Create testing set from 2025 data at specific sites
+testing <- all_matches_trimmed %>%
+  filter((year(sensor_datetime) == 2025 & id %in% c("chd", "sfm", "pbr", "pman", "pbd")))
+# Create training/validation set
+train_val <- all_matches_trimmed %>%
+  anti_join(testing)%>%
+  select(id, any_of(features), TOC_cat, sensor_datetime, collector)%>%
+  mutate(id = ifelse(id == "archery_virridy", "archery", id)) #combine archery and archery_virridy since they are the same site
```
Comment (steeleb) [inline modeling/toc/03_ross_only_categorical_TOC_model.Rmd]:
```suggestion
  mutate(id = if_else(id == "archery_virridy", "archery", id)) #combine archery and archery_virridy since they are the same site
```

File: modeling/toc/03_ross_only_categorical_TOC_model.Rmd [hunk 5]
```diff
@@ -0,0 +1,1281 @@
...
+    title = paste0(
+      "Training/Validation Set: ", nrow(train_val), " samples              ",
+      "Testing Set: ", nrow(testing), " samples"
+    ),
+    # subtitle = paste0(
+    #   "Training TOC Range: ", round(min(train_val$TOC, na.rm = TRUE), 2), " - ",
+    #   round(max(train_val$TOC, na.rm = TRUE), 2), " mg/L (Median: ", round(median(train_val$TOC), 2), ")\n",
+    #   "Testing TOC Range: ", round(min(testing$TOC, na.rm = TRUE), 2), " - ",
+    #   round(max(testing$TOC, na.rm = TRUE), 2), " mg/L (Median: ", round(median(testing$TOC), 2), ")"
+    # ),
```
Comment (steeleb) [inline modeling/toc/03_ross_only_categorical_TOC_model.Rmd]:
```suggestion
```

File: modeling/toc/03_ross_only_categorical_TOC_model.Rmd [hunk 6]
```diff
@@ -0,0 +1,1281 @@
...
+    # trimmed_features = c(general_features, "fdom_corr_max_6hr", "f_sc", "f_x_turb", "sin_doy" , "turb_median_12hr", "daily_canyon_flow_cfs")
+  ) , ~test_features(
+    train_val = train_val,
+    features_to_test = .x,
+    target = "TOC_cat",
+    default_hyper_params = default_hyper_params
+  ))
+```
+
+#### Feature Set Results
```
Comment (steeleb) [inline modeling/toc/03_ross_only_categorical_TOC_model.Rmd]: suggestion: can we reduce the number of TOC levels here? I think best practice is 3 or less.

File: modeling/toc/03_ross_only_categorical_TOC_model.Rmd [hunk 7]
```diff
@@ -0,0 +1,1281 @@
...
+#commented out so we don't overwrite existing model files
+# saveRDS(full_model, file = here("data", "derived", "models","TOC_realtime", "in_dev", paste0("toc_categorical_xgboost_models_",version,".rds")))
+#if we want to reload model
+#full_model <- readRDS(here("data", "derived", "models","TOC_realtime", "in_dev", paste0("toc_categorical_xgboost_models_",version,".rds")))
+
+rm(list = setdiff(ls(), c("train_val", "testing", "full_model","version","id_cols",
+                          "complete_sensor_data",
+                          #keeping functions around
+                          "apply_training_scale","plot_tvt_fold_acc","plot_train_test_ensemble_acc",
+                         "load_grab_sample_data", "plot_cat_model_timeseries" )) )
```
Comment (steeleb) [inline modeling/toc/03_ross_only_categorical_TOC_model.Rmd]:
```suggestion
                          "load_grab_sample_data", "plot_cat_model_timeseries")))
```

File: src/modeling/xgboost_hyperparameter_tuning_cat.R [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+      #train model with hyper parameters
+      model_ij <- xgb.train(
+        params = params,
+        data = dtrain,
+        nrounds = tune_grid$nrounds[j],
+        watchlist = watchlist,
+        #change early stopping rounds based on eta (smaller eta, larger rounds)
+        early_stopping_rounds = ifelse(tune_grid$eta[j] >= 0.1, 250,
+                                       ifelse(tune_grid$eta[j] >= 0.01, 500,
+                                              1000)),
```
Comment (steeleb) [inline src/modeling/xgboost_hyperparameter_tuning_cat.R]:
```suggestion
        early_stopping_rounds = if_else(tune_grid$eta[j] >= 0.1, 250,
                                       if_else(tune_grid$eta[j] >= 0.01, 500,
                                              1000)),
```

File: src/plotting/plot_train_test_ensemble_acc.R [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+
+    # --- Bind predictions back to data ---
+    train_df <- train_df %>%
+      mutate(!!pred_col := train_comb$pred_class,
+             !!paste0(pred_col, "_probs") := train_comb$max_prob)
+
+    test_df <- test_df %>%
+      mutate(!!pred_col := test_comb$pred_class,
+             !!paste0(pred_col, "_probs") := test_comb$max_prob)
+  }
```
Comment (steeleb) [inline src/plotting/plot_train_test_ensemble_acc.R]: flagging data.table syntax

File: src/plotting/plot_tvt_fold_acc.R [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+#'
+plot_tvt_fold_acc <- function(fold_model, fold_num,
+                              train_df = NULL, val_df = NULL, test_df = NULL,
+                              target_col = "TOC_cat", units = "mg/L", title = "",
+                              class_levels = c("0-2", "2-4", "4-8", "8+")) {
+
+  #See if there are any datasets
+  input_datasets <- list(train_df = train_df, val_df = val_df, test_df = test_df)
+
+  if (all(purrr::map_lgl(input_datasets, is.null))) {
```
Comment (steeleb) [inline src/plotting/plot_tvt_fold_acc.R]: flagging use of {package}:: syntax

**Review by steeleb — COMMENTED** [review]
Great push here!

I flagged some inconsistencies in syntax throughout, but really great extension of the flexibility of these functions and modeling efforts.

Generally, I advocate for not using global accuracy as a metric when reporting performance when there is unbalanced data (hence F1 and kappa which are better global metrics in unbalanced data sets). 

Question: Could we also start adding the knit output of documents like this one into the docs folder so that we don't have to run the workflow to think through the decisions? Basically, if decisions are being made, or it's a "frozen" model, we'll have a version of the knit markdown for reference?

File: src/modeling/xgboost_hyperparameter_tuning_cat.R [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+        print_every_n = 1000,
+        verbose = 0,
+        nthread = 1
+      )
+
+      pred_col <- paste0(target_col, "_guess")
+
+      # --- Validation Predictions ---
```
Comment (juandlt-csu) [inline src/modeling/xgboost_hyperparameter_tuning_cat.R]:
```suggestion
      # ---- Validation Predictions ----
```

Suggestion: I like that you are differentiating these sections by what they do. Big function with lots of things going on, this will make editing later (if necessary) easier. I suggest using at least four dashes so that R's outliner will pick up these sections automatically and we can jump to them easily. I will make a code suggestion to show what I mean.

File: src/modeling/xgboost_hyperparameter_tuning_cat.R
Comment (juandlt-csu) [inline src/modeling/xgboost_hyperparameter_tuning_cat.R]: For my own understanding, this function is meant for quick iterations on the model right? I am trying to get a sense of where I can go in to help y'all work "faster" and it seems that stuff like this for quick iteration would be good.

File: src/plotting/plot_cat_model_timeseries.R [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+#'
+#' @return A `ggplot` object showing predicted categorical classes over time
+#'   as colored tiles, with observed chemistry samples overlaid as black diamonds
+#'
+#' @examples
+#' \dontrun{
+#' plot_cat_model_timeseries(
+#'   model_input_df = sensor_data,
+#'   water_chem_df = chem_data,
+#'   models = map(full_model, "model),
```
Comment (juandlt-csu) [inline src/plotting/plot_cat_model_timeseries.R]:
```suggestion
#'   models = map(full_model, "model"),
```

File: src/modeling/xgboost_hyperparameter_tuning_cat.R (hunk 11 above)
Comment (SamStruthers) [reply src/modeling/xgboost_hyperparameter_tuning_cat.R]: Added!
